fe/fe-core/src/main/java/com/starrocks/scheduler/TaskManager.java

```diff
@@ -173,14 +240,294 @@ public ShowResultSet handleSubmitTaskStmt(SubmitTaskStmt submitTaskStmt) throws
+                taskRunManager.taskRunHistory.addHistory(taskRun.getStatus());
+            }
+        }
+    }
+
+    public void replayDropTaskRuns(List<String> queryIdList) {
+        Map<String, String> index = Maps.newHashMapWithExpectedSize(queryIdList.size());
+        for (String queryId : queryIdList) {
+            index.put(queryId, null);
+        }
+        taskRunManager.taskRunHistory.getAllHistory().removeIf(runStatus -> index.containsKey(runStatus.getQueryId()));
+    }
+
+    public void removeOldTaskInfo() {
+        long currentTimeMs = System.currentTimeMillis();
+
+        List<Task> currentTask = showTasks(null);
+        List<Long> taskIdToDelete = Lists.newArrayList();
+        currentTask.sort((o1, o2) -> Long.signum(o1.getCreateTime() - o2.getCreateTime()));
+        int labelKeepMaxSecond = Config.label_keep_max_second;
+        int numTaskToRemove = currentTask.size() - Config.label_keep_max_num;
+        for (Task task : currentTask) {
+            if ((currentTimeMs - task.getCreateTime()) / 1000 > labelKeepMaxSecond ||
+                    numTaskToRemove > 0) {
+                taskIdToDelete.add(task.getId());
+                --numTaskToRemove;
+            }
+        }
+        // this will do in checkpoint and does not need write log
+        dropTasks(taskIdToDelete, true);
+    }
+
+    public void removeOldTaskRunHistory() {
+        long currentTimeMs = System.currentTimeMillis();
+
+        Deque<TaskRunStatus> taskRunHistory = taskRunManager.taskRunHistory.getAllHistory();
+        List<String> historyToDelete = Lists.newArrayList();
+
+        if (!tryLock()) {
+            return;
+        }
+        try {
+            int labelKeepMaxSecond = Config.label_keep_max_second;
+            int numHistoryToRemove = taskRunHistory.size() - Config.label_keep_max_num;
+            for (TaskRunStatus runStatus : taskRunHistory) {
+                // task run may run for a long time and then failed
+                // there may be no finish time when it fails
+                long lastUpdateTime = runStatus.getCreateTime();
+                if (runStatus.getFinishTime() > lastUpdateTime) {
+                    lastUpdateTime = runStatus.getFinishTime();
+                }
+                if ((currentTimeMs - lastUpdateTime) / 1000 > labelKeepMaxSecond ||
+                        numHistoryToRemove > 0) {
+                    historyToDelete.add(runStatus.getQueryId());
+                    taskRunHistory.remove();
+                    --numHistoryToRemove;
+                }
+            }
+        } finally {
+            unlock();
+        }
+        LOG.info("remove run history:{}", historyToDelete);
+    }
+
+
+    private static class SerializeData {
+        @SerializedName("tasks")
+        public List<Task> tasks;
+
+        @SerializedName("runStatus")
+        public List<TaskRunStatus> runStatus;
+    }
+
+    static class TaskRunManager {
```

Comment:
Why not write TaskRunManager in a specific class.

Reply:
Before, it was because the pending and running taskRuns needed to be set to Failed when the FE restarted, so the TaskRunManager needed to access the manualTaskMap field of the TaskManager, but after the refactoring, it seems that getting the relevant methods to the TaskManager does not require the use of inner classes So I will use a separate class recover the code.